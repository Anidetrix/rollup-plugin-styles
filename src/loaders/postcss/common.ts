export const hashRe = /\[hash(?::(\d+))?]/;
export const firstExtRe = /(\.[^\s.]+)/i;
export const dataURIRe = /data:[^\n\r;]+?(?:;charset=[^\n\r;]+?)?;base64,([\d+/A-Za-z]+={0,2})/;
