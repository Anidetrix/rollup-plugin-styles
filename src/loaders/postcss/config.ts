import path from "path";
import postcss from "postcss";
import { cosmiconfig } from "cosmiconfig";
import { PostCSSConfigLoaderOptions } from "../../types";
import { ensurePCSSPlugins, ensurePCSSOption } from "../../utils/options";

interface Options {
  parser?: postcss.Parser;
  syntax?: postcss.Syntax;
  stringifier?: postcss.Stringifier;
}

interface Config extends Options {
  plugins?: { [p: string]: Record<string, unknown> } | (string | postcss.Plugin<unknown>)[];
}

interface Result {
  plugins: (postcss.Transformer | postcss.Processor)[];
  options: Options;
}

export default async function (
  id: string,
  config: false | PostCSSConfigLoaderOptions,
): Promise<Result> {
  if (!config) return { plugins: [], options: {} };

  const { ext, dir, base } = path.parse(id);

  type Found = { config: Config | ((ctx: Record<string, unknown>) => Config); isEmpty?: boolean };
  const searchPath = config.path ? path.resolve(config.path) : dir;
  const found: Found | null = await cosmiconfig("postcss").search(searchPath);

  if (!found || found.isEmpty) return { plugins: [], options: {} };

  const { plugins, parser, syntax, stringifier } =
    typeof found.config === "function"
      ? found.config({
          cwd: process.cwd(),
          env: process.env["NODE_ENV"] ?? "development",
          file: { extname: ext, dirname: dir, basename: base },
          options: config.ctx ?? {},
        })
      : found.config;

  const result: Result = { plugins: ensurePCSSPlugins(plugins), options: {} };
  if (parser) result.options.parser = ensurePCSSOption(parser, "parser");
  if (syntax) result.options.syntax = ensurePCSSOption(syntax, "syntax");
  if (stringifier) result.options.stringifier = ensurePCSSOption(stringifier, "stringifier");

  return result;
}
