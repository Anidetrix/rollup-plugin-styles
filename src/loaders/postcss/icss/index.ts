import postcss from "postcss";
import { extractICSS, replaceSymbols, replaceValueSymbols, Replacements } from "icss-utils";

import loadDefault, { Load } from "./load";
import resolve from "./resolve";

const name = "styles-icss";

export type InteroperableCSSOptions = {
  load?: Load;
  getReplacements?: (file: string, replacements: Replacements, out?: string) => void;
  extensions?: string[];
};

const plugin: postcss.Plugin<InteroperableCSSOptions> = postcss.plugin(
  name,
  (options = {}) => async (css, res): Promise<void> => {
    if (!css.source?.input.file) return;
    if (!res.processor) return;

    const load = options.load ?? loadDefault;
    const extensions = options.extensions ?? [".css", ".pcss", ".postcss", ".sss"];

    const opts = res.opts && { ...res.opts };
    delete opts?.map;

    const { icssExports, icssImports } = extractICSS(css);

    const imports = await resolve(
      icssImports,
      load,
      css.source.input.file,
      extensions,
      res.processor,
      opts,
    );

    replaceSymbols(css, imports);

    const replacements = Object.entries(icssExports).reduce<Replacements>((acc, [k, v]) => {
      acc[k] = replaceValueSymbols(v, imports);
      return acc;
    }, {});

    res.messages.push({ plugin: name, type: "icss", replacements });

    if (typeof options.getReplacements === "function")
      options.getReplacements(css.source.input.file, replacements, res.opts?.to);
  },
);

export default plugin;
