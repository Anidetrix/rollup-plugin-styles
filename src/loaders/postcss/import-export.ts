import path from "path";
import fs from "fs-extra";
import { AsyncOpts } from "resolve";
import postcss from "postcss";
import { extractICSS, replaceSymbols, replaceValueSymbols, CSSImports } from "icss-utils";

import resolveAsync from "../../utils/resolve-async";
import { normalizePath } from "../../utils/path-utils";

const name = "postcss-import-export-plugin";

type Json = { [k: string]: string };

/** Type of function for loading ICSS from composed files */
export type Load = (
  url: string,
  file: string,
  extensions: string[],
  processor: postcss.Processor,
) => Promise<Json>;

export interface ImportExportOptions {
  load?: Load;
  getJSON?: (file: string, json: Json, out?: string) => void;
  extensions?: string[];
}

const defaultLoad: Load = async (url, file, extensions, processor) => {
  let from: string;
  const properUrl = normalizePath(url);
  const asyncOpts: AsyncOpts = { basedir: path.dirname(file), extensions };
  try {
    from = await resolveAsync(properUrl, asyncOpts);
  } catch (error) {
    from = await resolveAsync(`./${properUrl}`, asyncOpts);
  }
  const source = await fs.readFile(from, "utf8");
  const { messages } = await processor.process(source, { from });
  return messages
    .filter(msg => msg.plugin === name && msg.type === "export")
    .map(msg => msg.json as Json)
    .reduce((prev, current) => ({ ...prev, ...current }));
};

/**
 * @param icssImports ICSS Imports
 * @param load Dependency loader
 * @param file Input file
 * @param extensions File extensions
 * @param processor PostCSS Processor
 * @returns Resolved imported values
 */
function resolveImportedValues(
  icssImports: CSSImports,
  load: Load,
  file: string,
  extensions: string[],
  processor: postcss.Processor,
): Promise<Json> {
  return Object.entries(icssImports).reduce(async (result, [url, values]) => {
    const exports = await load(url, file, extensions, processor);
    const mappedValues = Object.entries(values).reduce((acc, [k, v]) => {
      acc[k] = exports[v];
      return acc;
    }, {} as Json);
    return {
      ...(await result),
      ...mappedValues,
    };
  }, Promise.resolve({}));
}

const plugin: postcss.Plugin<ImportExportOptions> = postcss.plugin(
  name,
  options => async (css, res): Promise<void> => {
    if (!css.source) throw new Error("No source detected!");
    if (!css.source.input.file) throw new Error("`from` field required!");
    if (!res.processor) throw new Error("No processor detected!");

    const load = (options && options.load) || defaultLoad;
    const extensions = (options && options.extensions) || [".css"];

    const { icssExports, icssImports } = extractICSS(css);

    const importedValues = await resolveImportedValues(
      icssImports,
      load,
      css.source.input.file,
      extensions,
      res.processor,
    );

    replaceSymbols(css, importedValues);

    const json = Object.entries(icssExports).reduce((acc, [k, v]) => {
      acc[k] = replaceValueSymbols(v, importedValues);
      return acc;
    }, {} as Json);

    // Needed mainly for the plugin itself, for recursion
    res.messages.push({ plugin: name, type: "export", json });

    if (options && typeof options.getJSON === "function")
      options.getJSON(css.source.input.file, json, res.opts && res.opts.to);
  },
);

export default plugin;
