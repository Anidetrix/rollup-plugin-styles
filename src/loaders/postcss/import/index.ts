import path from "path";
import postcss from "postcss";
import valueParser from "postcss-value-parser";

import { normalizePath } from "../../../utils/path";

import resolveDefault, { Resolve } from "./resolve";

const name = "styles-import";

/** `@import` handler options */
export type ImportOptions = {
  /**
   * Provide custom resolver for imports
   * in place of the default one
   */
  resolve?: Resolve;
  /**
   * Aliases for import paths.
   * Overrides the global `alias` option.
   * - ex.: `{"foo":"bar"}`
   */
  alias?: { [from: string]: string };
};

type ImportPrivateOptions = { extensions?: string[] };
const plugin: postcss.Plugin<ImportOptions & ImportPrivateOptions> = postcss.plugin(
  name,
  (options = {}) => async (css, res): Promise<void> => {
    if (!css.source?.input.file) return;

    const resolve = options.resolve ?? resolveDefault;
    const alias = options.alias ?? {};
    const extensions = options.extensions ?? [".css", ".pcss", ".postcss", ".sss"];

    const opts = res.opts && { ...res.opts };
    delete opts?.map;

    const { file } = css.source.input;
    const importSet = new Set<{ importRule: postcss.AtRule; url: string }>();
    const basedir = path.dirname(file);

    css.walkAtRules(/^import$/i, importRule => {
      // Top level only
      if (importRule.parent.type !== "root") {
        importRule.warn(res, "`@import` should be top level");
        return;
      }

      // Child nodes should not exist
      if (importRule.nodes) {
        importRule.warn(res, "`@import` was not terminated correctly");
        return;
      }

      const [urlNode] = valueParser(importRule.params).nodes;

      // No URL detected
      if (!urlNode || (urlNode.type !== "string" && urlNode.type !== "function")) {
        importRule.warn(res, `No URL in \`${importRule.toString()}\``);
        return;
      }

      let url = "";

      if (urlNode.type === "string") {
        url = urlNode.value;
      } else if (urlNode.type === "function") {
        // Invalid function
        if (urlNode.value.toLowerCase() !== "url") {
          importRule.warn(res, `Invalid \`url\` function in \`${importRule.toString()}\``);
          return;
        }

        const isString = urlNode.nodes[0]?.type === "string";
        url = isString ? urlNode.nodes[0].value : valueParser.stringify(urlNode.nodes);
      }

      url = url.replace(/^\s+|\s+$/g, "");

      // Resolve aliases
      for (const [from, to] of Object.entries(alias)) {
        if (!url.startsWith(from)) continue;
        url = normalizePath(to) + url.slice(from.length);
      }

      // Empty url
      if (url.length === 0) {
        importRule.warn(res, `Empty URL in \`${importRule.toString()}\``);
        return;
      }

      importSet.add({ importRule, url });
    });

    for await (const { importRule, url } of importSet) {
      try {
        const { source, from } = await resolve(url, basedir, extensions);

        if (!(source instanceof Uint8Array) || typeof from !== "string") {
          importRule.warn(res, `Incorrectly resolved \`@import\` in \`${importRule.toString()}\``);
          continue;
        }

        if (normalizePath(from) === normalizePath(file)) {
          importRule.warn(res, `\`@import\` loop in \`${importRule.toString()}\``);
          continue;
        }

        res.messages.push({ plugin: name, type: "dependency", file: from });

        const imported = await postcss(plugin(options)).process(source, { ...opts, from });
        res.messages.unshift(...imported.messages);

        if (!imported.root) importRule.remove();
        else importRule.replaceWith(imported.root);
      } catch {
        importRule.warn(res, `Unresolved \`@import\` in \`${importRule.toString()}\``);
      }
    }
  },
);

export default plugin;
