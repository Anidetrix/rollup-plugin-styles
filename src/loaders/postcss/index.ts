import path from "path";
import { RawSourceMap } from "source-map";
import { makeLegalIdentifier } from "@rollup/pluginutils";
import postcss from "postcss";
import loadPostCSSConfig from "postcss-load-config";
import cssnano from "cssnano";

import { Loader, PostCSSLoaderOptions } from "../../types";
import { humanlizePath, normalizePath } from "../../utils/path";
import { mm } from "../../utils/sourcemap";
import resolveAsync from "../../utils/resolve-async";
import safeId from "../../utils/safe-id";
import { denullifyObject, booleanFilter } from "../../utils/filter";

import postcssImport from "./import";
import postcssUrl from "./url";
import postcssModules from "./modules";
import postcssICSS from "./icss";
import postcssNoop from "./noop";

type LoadedConfig = ReturnType<typeof loadPostCSSConfig> extends PromiseLike<infer T> ? T : never;

/**
 * @param id File path
 * @param config `postcss-load-config`'s options
 * @returns Loaded PostCSS config
 */
async function loadConfig(
  id: string,
  config: PostCSSLoaderOptions["config"],
): Promise<Partial<LoadedConfig>> {
  const configPath =
    typeof config === "object" && config.path ? path.resolve(config.path) : path.dirname(id);

  const context: { [prop: string]: object | string } = {
    file: {
      extname: path.extname(id),
      dirname: path.dirname(id),
      basename: path.basename(id),
    },
    options: typeof config === "object" ? config.ctx ?? {} : {},
  };

  return loadPostCSSConfig(context, configPath).catch(error => {
    if (!(error.message as string).toLowerCase().includes("no postcss config found")) throw error;
    return {};
  });
}

const loader: Loader<PostCSSLoaderOptions> = {
  name: "postcss",
  alwaysProcess: true,
  async process({ code, map, extracted }) {
    const { options } = this;

    const config = await loadConfig(this.id, options.config);

    const plugins = [
      ...[
        options.import && postcssImport({ extensions: options.extensions, ...options.import }),
        options.url && postcssUrl(options.url),
      ].filter(booleanFilter),
      ...(options.postcss.plugins ?? []),
      ...(config.plugins ?? []),
    ];

    const autoModules =
      options.autoModules &&
      (typeof options.autoModules === "function"
        ? options.autoModules(this.id)
        : options.autoModules instanceof RegExp
        ? options.autoModules.test(this.id)
        : /\.module\.[A-Za-z]+$/.test(this.id));

    const supportModules = Boolean(options.modules || autoModules);

    const modulesExports: { [filepath: string]: { [prop: string]: string } } = {};

    const postcssOpts: PostCSSLoaderOptions["postcss"] & {
      from: string;
      to: string;
      map: postcss.ProcessOptions["map"];
    } = {
      ...denullifyObject((config.options ?? {}) as Required<postcss.ProcessOptions>),
      ...denullifyObject(options.postcss),
      from: this.id,
      to: options.to ?? this.id,
      map: {
        inline: false,
        annotation: false,
        sourcesContent: true,
        prev: mm(map).relative(path.dirname(this.id)).toObject(),
      },
    };

    delete postcssOpts.plugins;

    if (supportModules) {
      const modulesOptions = typeof options.modules === "object" ? options.modules : {};
      plugins.push(
        ...postcssModules({
          // Skip hash while testing since CSS content would differ on Windows and Linux
          // due to different line endings.
          generateScopedName: process.env.NODE_ENV === "test" ? "[name]_[local]" : undefined,
          failOnWrongOrder: true,
          ...modulesOptions,
        }),
        postcssICSS({
          extensions: options.extensions,
          getReplacements(file, replacements, out) {
            modulesExports[file] = replacements;
            if (
              typeof options.modules === "object" &&
              typeof options.modules.getReplacements === "function"
            ) {
              return options.modules.getReplacements(file, replacements, out);
            }
          },
        }),
      );
    }

    if (options.minimize)
      plugins.push(cssnano(typeof options.minimize === "object" ? options.minimize : {}));

    // Avoid PostCSS warning
    if (plugins.length === 0) plugins.push(postcssNoop);

    const res = await postcss(plugins).process(code, postcssOpts);

    for (const warning of res.warnings())
      this.warn({ name: warning.plugin, message: warning.text });

    const deps = res.messages.filter(msg => msg.type === "dependency");
    for (const dep of deps) this.deps.add(normalizePath(dep.file));

    const assets = res.messages.filter(msg => msg.type === "asset");
    for (const asset of assets) this.assets.set(asset.to, asset.source);

    map = mm((res.map?.toJSON() as unknown) as RawSourceMap)
      .resolve(path.dirname(postcssOpts.to))
      .toString();

    if (!options.extract && this.sourceMap)
      res.css += mm(map)
        .modify(map => void delete map.file)
        .relative()
        .toCommentData();

    if (options.emit) return { code: res.css, map };

    const reservedWords = new Set<string>();
    const saferId = (id: string): string => safeId(id, humanlizePath(this.id));

    const cssExportName = "css";
    reservedWords.add(cssExportName);

    const cssVarName = saferId(cssExportName);
    const modulesVarName = saferId("modules");

    const output = [
      `const ${cssVarName} = ${JSON.stringify(res.css)}`,
      `const ${modulesVarName} = ${JSON.stringify(modulesExports[this.id] ?? {})}`,
      `export const ${cssExportName} = ${cssVarName}`,
      `export default ${supportModules ? modulesVarName : cssVarName}`,
    ];

    if (options.namedExports) {
      const json = modulesExports[this.id];

      const getClassName =
        typeof options.namedExports === "function"
          ? options.namedExports
          : (name: string): string =>
              makeLegalIdentifier(reservedWords.has(name) ? `_${name}` : name);

      for (const name in json) {
        const newName = getClassName(name);

        if (name !== newName)
          this.warn(`Exported \`${name}\` as \`${newName}\` in ${humanlizePath(this.id)}`);

        if (!json[newName]) json[newName] = json[name];

        output.push(`export const ${newName} = ${JSON.stringify(json[name])}`);
      }
    }

    if (options.extract) extracted = { id: this.id, css: res.css, map };

    if (options.inject) {
      if (typeof options.inject === "function") {
        output.push(options.inject(cssVarName, this.id));
      } else {
        const injectorVarName = saferId("injector");
        const injectorId = await resolveAsync("./inject-css", {
          basedir: path.join(
            process.env.NODE_ENV === "test" ? process.cwd() : __dirname,
            "runtime",
          ),
        });
        const injectorData =
          typeof options.inject === "object" ? `,${JSON.stringify(options.inject)}` : "";

        output.push(
          `import ${injectorVarName} from '${normalizePath(injectorId)}'`,
          `${injectorVarName}(${cssVarName}${injectorData})`,
        );
      }
    }

    code = `${output.join(";\n")};`;

    return { code, map, extracted };
  },
};

export default loader;
