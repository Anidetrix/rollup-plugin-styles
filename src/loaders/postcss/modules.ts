import { makeLegalIdentifier } from "@rollup/pluginutils";

import path from "path";
import { Plugin as PostCSSPlugin } from "postcss";
import modulesValues from "postcss-modules-values";
import localByDefault from "postcss-modules-local-by-default";
import extractImports from "postcss-modules-extract-imports";
import modulesScope, { ScopeOptions } from "postcss-modules-scope";

import { ModulesOptions } from "../../types";
import { humanlizePath } from "../../utils/path-utils";
import hash from "../../utils/hash";

const hashRe = /\[hash(?::(\d+))?]/;

type HashData = { name: string; filename: string; css: string };
/**
 * @param {object} data Hash Data
 * @returns {string} Hash
 */
function getHash({ name, filename, css }: HashData): string {
  return hash(`${humanlizePath(filename)};${name};${css}`);
}

/**
 * @param {string|undefined} placeholder string with placeholders
 * @returns {Function} function for generating scoped name
 */
function getGenerator(placeholder?: string): NonNullable<ScopeOptions["generateScopedName"]> {
  if (placeholder) {
    return (name: string, filename: string, css: string): string => {
      const hash = getHash({ name, filename, css });
      const match = hashRe.exec(placeholder);
      const hashLen = match && parseInt(match[1]);
      return makeLegalIdentifier(
        placeholder
          .replace("[name]", path.basename(filename, path.extname(filename)))
          .replace("[local]", name)
          .replace(hashRe, hashLen ? hash.slice(0, hashLen) : hash),
      );
    };
  }

  return (name: string, filename: string, css: string): string => {
    const i = css.indexOf(`.${name}`);
    const lineNumber = css.slice(0, i).split(/[\n\r]/).length;
    const hash = getHash({ name, filename, css }).slice(0, 8);
    return makeLegalIdentifier(`${name}_${hash}_${lineNumber}`);
  };
}

export default (options: ModulesOptions): PostCSSPlugin<unknown>[] => {
  const opts = {
    mode: "local" as "local",
    ...options,
    generateScopedName:
      typeof options.generateScopedName === "function"
        ? options.generateScopedName
        : getGenerator(options.generateScopedName),
  };

  return [
    modulesValues(),
    localByDefault({ mode: opts.mode }),
    extractImports({ failOnWrongOrder: opts.failOnWrongOrder }),
    modulesScope({
      exportGlobals: opts.exportGlobals,
      generateScopedName: opts.generateScopedName,
    }),
  ];
};
