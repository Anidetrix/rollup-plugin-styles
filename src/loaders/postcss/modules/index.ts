import postcss from "postcss";
import modulesValues from "postcss-modules-values";
import localByDefault from "postcss-modules-local-by-default";
import extractImports from "postcss-modules-extract-imports";
import modulesScope from "postcss-modules-scope";
import generateScopedNameDefault from "./generate";

/** Options for [CSS Modules](https://github.com/css-modules/css-modules) */
export interface ModulesOptions {
  /**
   * Default mode for classes
   * @default "local"
   */
  mode?: "local" | "global" | "pure";
  /** Fail on wrong order of composition */
  failOnWrongOrder?: boolean;
  /** Export global classes */
  exportGlobals?: boolean;
  /**
   * Placeholder or function for scoped name generation.
   * Allowed blocks for placeholder:
   * - `[dir]`: The directory name of the asset.
   * - `[name]`: The file name of the asset excluding any extension.
   * - `[local]`: The original value of the selector.
   * - `[hash(:<num>)]`: A hash based on the name and content of the asset (with optional length).
   * @default "[name]_[local]__[hash:8]"
   */
  generateScopedName?: string | ((name: string, file: string, css: string) => string);
}

export default (options: ModulesOptions): (postcss.Transformer | postcss.Processor)[] => {
  const opts = {
    mode: "local" as const,
    ...options,
    generateScopedName:
      typeof options.generateScopedName === "function"
        ? options.generateScopedName
        : generateScopedNameDefault(options.generateScopedName),
  };

  return [
    modulesValues(),
    localByDefault({ mode: opts.mode }),
    extractImports({ failOnWrongOrder: opts.failOnWrongOrder }),
    modulesScope({
      exportGlobals: opts.exportGlobals,
      generateScopedName: opts.generateScopedName,
    }),
  ];
};
