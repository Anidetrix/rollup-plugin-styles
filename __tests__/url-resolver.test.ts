import postcss from "postcss";
import urlResolver, { UrlOptions } from "../src/loaders/postcss/url";
// import { fixture } from "./helpers";

const validateUrl = async (
  css: string,
  options: UrlOptions = {},
  from = "dummy",
): Promise<string> => {
  const data = await postcss(urlResolver(options)).process(css, { from });
  const [warning] = data.warnings();
  return warning.text;
};

describe("url resolver", () => {
  it("warns about being empty", async () => {
    const warning = await validateUrl(".foo{background:url()}");
    expect(warning).toMatchSnapshot("warning");
  });

  it("warns about being unresolved", async () => {
    const warning = await validateUrl(".foo{background:url(bg.png)}");
    expect(warning).toMatchSnapshot("warning");
  });

  it("warns about incorrect resolving", async () => {
    const warning = await validateUrl(".foo{background:url(bg.png)}", {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any,@typescript-eslint/no-unsafe-return
      resolve: () => "lol" as any,
    });
    expect(warning).toMatchSnapshot("warning");
  });
});
