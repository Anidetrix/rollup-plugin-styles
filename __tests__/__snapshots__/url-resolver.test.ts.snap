// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url resolver warns about being empty: warning 1`] = `"Empty URL in \`background:url()\`"`;

exports[`url resolver warns about being unresolved: warning 1`] = `"Unresolved URL \`bg.png\` in \`background:url(bg.png)\`"`;

exports[`url resolver warns about incorrect resolving: warning 1`] = `"Incorrectly resolved URL \`bg.png\` in \`background:url(bg.png)\`"`;
